### DISCUSSION TITLE
Common definition for class, interface and object declaration

### DISCUSSION THREAD

**Discussion by andriishpek-sigma** [LemonAppDev/konsist#1027] [Ask a Question] [answered]
### Wanted end result.

Is there existing way to pick all classes, interfaces and objects from scope? They have a lot in common: names, children etc...

### What you tried so far.

I'm using list conjunction for now: `classes() + objects()`. Although it works only if used directly in test: Kotlin automatically infers common types. But it doesn't work when moved to extension function:
```
fun KoScope.classesAndObjects() = classes() + objects()
```

**Comment by igorwojda** [comment]
We have been thinking about this, but the initial spike have resulted in many edge cases.
I see that this feature would be handy, so we will get back to this at some point.

**Reply by andriishpek-sigma** [reply]
Is it possible just to declare an aggregate interface/contract with all common types, and add to those three declarations? If you haven't tried this yet, but think it may work, then I can give it a try and set up a PR

**Reply by igorwojda** [reply]
@andriishpek-sigma Please give it a go.
BTW please read this guide https://docs.konsist.lemonappdev.com/help/contributing

**Comment by igorwojda** [comment]
BTW 
@andriishpek-sigma 
It is also possible to filter declarations by using certain aspects e.g. visibility modifiers. Usage of `providers` allows verifying the visibility of different declaration types such as classes, functions, properties, etc:

```kotlin
koScope
    .declarationsOf<KoVisibilityModifierProvider>()
    .assertTrue { it.hasInternalModifier }
```

**Reply by andriishpek-sigma** [reply]
It's going to work indeed, but
1. you are limited to accessing single `Provider` properties
2. you'll have to resolve scope declarations every time you need a different `Provider`; on my side it's slow in some cases (`indirectParents`), but now it'll skyrocket (upside down) with all other minor declarations

**Reply by igorwojda** [reply]
I agree this is not ideal. 
Can you provide example of a concrete test you would like to write?

**Reply by andriishpek-sigma** [reply]
Simple example:

```
private val scope = Konsist.scopeFromAppModule()
private val actionsClasses = pickActions { classes() }
private val actionsInterfaces = pickActions { interfaces() }
private val actionsObjects = pickActions { objects() }

private fun <T : KoParentInterfaceProvider> pickActions(
    provider: KoScope.() -> List<T>,
): List<T> = scope
    .provider()
    .withParentInterfaceOf(ReduxAction::class, indirectParents = true)

@Test
fun `All action declarations are public`() {
    run { actionsClasses + actionsObjects }
        .withoutPackage(ReduxArchTest.CORE_PACKAGE)
        .assertTrue {
            it.hasPublicOrDefaultModifier
        }
}
```

**Comment by nataliapeterwas** [comment, marked as answer]
This improvement will be added in the next release: [PR](https://github.com/LemonAppDev/konsist/pull/1260) :)

**Comment by nataliapeterwas** [comment]
Konsist `0.16.0` has been released! This release includes this improvement :)
